core/nixpkgs.bzl

```diff
@@ -668,3 +670,224 @@ def nixpkgs_package(
         kwargs["repositories"] = inversed_repositories
 
     _nixpkgs_package(**kwargs)
+
+def _nixpkgs_flake_package_impl(repository_ctx):
+    # Workaround to bazelbuild/bazel#4533 -- to prevent this rule being restarted after running cp,
+    # resolve all dependencies of this rule before running cp
+    #
+    # Remove the following repository_ctx.path() once bazelbuild/bazel#4533 is resolved.
+    if repository_ctx.attr.build_file:
+        repository_ctx.path(repository_ctx.attr.build_file)
+
+    repository_ctx.path(repository_ctx.attr.nix_flake_file)
+    repository_ctx.path(external_repository_root(repository_ctx.attr.nix_flake_file))
+    repository_ctx.path(repository_ctx.attr.nix_flake_lock_file)
+    repository_ctx.path(external_repository_root(repository_ctx.attr.nix_flake_lock_file))
+
+    for dep in repository_ctx.attr.nix_flake_file_deps:
+        repository_ctx.path(dep)
+        repository_ctx.path(external_repository_root(dep))
+
+    # Is nix supported on this platform?
+    not_supported = not is_supported_platform(repository_ctx)
+
+    # Should we fail if Nix is not supported?
+    fail_not_supported = repository_ctx.attr.fail_not_supported
+
+    if not_supported and fail_not_supported:
+        fail("Platform is not supported: `nix` not found in PATH. See attribute `fail_not_supported` if you don't want to use Nix.")
+    elif not_supported:
+        return
+
+    # If true, a BUILD file will be created from a template if it does not
+    # exist.
+    # However this will happen AFTER the nix-build command.
+    create_build_file_if_needed = False
+    if repository_ctx.attr.build_file and repository_ctx.attr.build_file_content:
+        fail("Specify one of 'build_file' or 'build_file_content', but not both.")
+    elif repository_ctx.attr.build_file:
+        repository_ctx.symlink(repository_ctx.attr.build_file, "BUILD")
+    elif repository_ctx.attr.build_file_content:
+        repository_ctx.file("BUILD", content = repository_ctx.attr.build_file_content)
+    else:
+        # No user supplied build file, we may create the default one.
+        create_build_file_if_needed = True
+
+        # Workaround to bazelbuild/bazel#4533
+        repository_ctx.path("BUILD")
+
+    nix_flake_file_deps = {}
+    for dep_lbl, dep_str in repository_ctx.attr.nix_flake_file_deps.items():
+        nix_flake_file_deps[dep_str] = cp(repository_ctx, dep_lbl)
```

Comment:
This could potentially be grouped with the `expand_location()` call below and pulled out to be shared with `nixpkgs_package()`.